tests/unit/wordpress_mock.py

```diff
@@ -0,0 +1,526 @@
+
+        Args:
+            match: A match function takes input and output matching result as bool.
+        """
+
+        def decorator(func):
+            self.registered_handler.append((match, func))
+            return func
+
+        return decorator
+
+
+class ExecProcessMock:
+    """A mock for :class:`ops.pebble.ExecProcess`."""
+
+    def __init__(self, return_code: int, stdout: str, stderr: str):
+        self._return_code = return_code
+        self._stdout = stdout
+        self._stderr = stderr
+
+    def wait_output(self) -> typing.Tuple[str, str]:
+        if self._return_code != 0:
+            raise ops.pebble.ExecError(
+                [], exit_code=self._return_code, stdout=self._stdout, stderr=self._stderr
+            )
+        return self._stdout, self._stderr
+
+
+class WordpressContainerMock:
+    """A mock for :class:`ops.charm.model.Container`.
+
+    This will simulate file system and subprocess system inside the WordPress container.
+    """
+
+    _exec_handler = HandlerRegistry()
+
+    def __init__(
+        self,
+        wordpress_database_mock: WordpressDatabaseMock,
+    ):
+        self.original_pebble = None
+        self.fs: typing.Dict[str, str] = {}
+        self._wordpress_database_mock = wordpress_database_mock
+        self.installed_plugins = set(WordpressCharm._WORDPRESS_DEFAULT_PLUGINS)
+        self.installed_themes = set(WordpressCharm._WORDPRESS_DEFAULT_THEMES)
+        self.wp_eval_history = []
+
+    def exec(
+        self, cmd, user=None, group=None, working_dir=None, combine_stderr=None, timeout=None
+    ):
+        handler = None
+        for match, potential_handler in self._exec_handler.registered_handler:
+            is_match = match(cmd)
+            if is_match and handler is not None:
+                raise ValueError(f"Multiple handlers registered for the same cmd {cmd}")
+            if is_match:
+                handler = potential_handler
+        if handler is None:
+            raise ValueError(f"No handler registered for the cmd {cmd}")
+        return handler(self, cmd)
+
+    def pull(self, path: str) -> typing.IO[str]:
+        return io.StringIO(self.fs[path])
+
+    def push(self, path: str, source: str, user=None, group=None, permissions=None) -> None:
+        self.fs[path] = source
+
+    def exists(self, path):
+        return path in self.fs
+
+    def list_files(self, path: str) -> typing.List[str]:
+        if not path.endswith("/"):
+            path += "/"
+        file_list = []
+        for file in self.fs:
+            if file.startswith(path):
+                file_list.append(file.removeprefix(path).split("/")[0])
+        return file_list
+
+    def remove_path(self, path: str, recursive: bool = False) -> None:
+        try:
+            del self.fs[path]
+        except KeyError:
+            if recursive:
+                return
+            raise
+
+    def _get_current_database_config(self):
```

Comment:
For this and below, should this have a return type?

Reply:
Fixed, thanks!